gigascience/gigadb-website
# New feature/1632 replace attribute text field with a drop down menu

# Pull request for issue: #1632 #1611

This PR should be merged after https://github.com/gigascience/gigadb-website/pull/1747 is merged

This is a pull request for the following functionalities:

* Show list of selectable attributes in metadata editor
![image](https://github.com/gigascience/gigadb-website/assets/143437854/70b80e9c-4dad-4b5d-8ac4-706b91f23448)

## How to test?

* Create file `files/data/attributes.json` with content:
```
[
  {
    "attribute_name": "age",
    "definition": "age of host at the time of sampling; relevant scale depends on species and study, e.g. could be seconds for amoebae or centuries for trees",
    "allowed_units": ["UO:0000031","UO:0000032","UO:0000033","UO:0000034","UO:0000035","UO:0000036"],
    "attachable_to_files": true,
    "attachable_to_samples": true,
    "attachable_to_datasets": false
  },
  {
    "attribute_name": "sex",
    "definition": "physical sex of the host",
    "allowed_units": ["male","female","neuter","hermaphrodite","not determined"],
    "attachable_to_files": true,
    "attachable_to_samples": true,
    "attachable_to_datasets": false
  },
  {
    "attribute_name": "height",
    "definition": "height of the host",
    "allowed_units": [],
    "attachable_to_files": false,
    "attachable_to_samples": true,
    "attachable_to_datasets": false
  }
]
```
* go to annotate files, i.e. http://gigadb.gigasciencejournal.com/authorisedDataset/annotateFiles/id/200070
* verify that there is a dorpdown in the drawer with age and sex but not height

## How have functionalities been implemented?

* the content of `files/data/attributes.json` is expected to be valid JSON and passed as prop to the vue widget that renders the attribute selector
* only the attribute name is considered, other fields such as allowed units are ignored
* only the attributes which have a truthy `attachable_to_files` value are considered
* Added a dynamically displayed definition text for the currently selected attribute

Please, consider this feature a draft. Dynamically showing the definition might not be the best user experience because the user needs to first select an attribute to read the definition. I think some other solution would be preferrable

A more complex solution would be to create a custom made select widget, instead of using the native html. This would allow:
* to show the description alongside the label in a way that is visually okay , e.g.
![image](https://github.com/gigascience/gigadb-website/assets/143437854/8089ca85-6a0f-4903-9b08-73645dae6206)
* if there are many (hundreds) attributes to display, fuzzy search and inifinite scrolling or virtualization might improve user experience and performance

A simple alternative would be to add a static list with attributes and definitions somewhere in the guidelines pages, and link to it from the description of the attributes dropdown. It might depend on whether the users need to see a definition of each attribute every time, or we can assume that they are going to be familiarized with the attributes

Based on feedback I can apply one or the other

## Any issues with implementation?

--

## Any changes to automated tests?

--
